SignalServiceKit/src/Profiles/OWSProfilesManager.m

```diff
@@ -375,32 +436,164 @@ - (void)updateProfileOnService:(nullable NSString *)localProfileName
     });
 }
 
-- (void)loadLocalProfileAsync
+#pragma mark - Profile Whitelist
+
+- (void)addUserToProfileWhitelist:(NSString *)recipientId
+{
+    OWSAssert(recipientId.length > 0);
+
+    [self.dbConnection setObject:@(1) forKey:recipientId inCollection:kOWSProfilesManager_UserWhitelistCollection];
+    self.userProfileWhitelistCache[recipientId] = @(YES);
+}
+
+- (BOOL)isUserInProfileWhitelist:(NSString *)recipientId
 {
-    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-        NSString *_Nullable localProfileName = [self.dbConnection objectForKey:kOWSProfilesManager_LocalProfileNameKey
-                                                                  inCollection:kOWSProfilesManager_Collection];
-        AvatarMetadata *_Nullable localProfileAvatarMetadata =
-            [self.dbConnection objectForKey:kOWSProfilesManager_LocalProfileAvatarMetadataKey
-                               inCollection:kOWSProfilesManager_Collection];
-        UIImage *_Nullable localProfileAvatarImage = nil;
-        if (localProfileAvatarMetadata) {
-            localProfileAvatarImage = [self loadProfileAvatarWithFilename:localProfileAvatarMetadata.fileName];
-            if (!localProfileAvatarImage) {
-                localProfileAvatarMetadata = nil;
-            }
+    OWSAssert(recipientId.length > 0);
+
+    NSNumber *_Nullable value = self.userProfileWhitelistCache[recipientId];
+    if (value) {
+        return [value boolValue];
+    }
+
+    value =
+        @(nil != [self.dbConnection objectForKey:recipientId inCollection:kOWSProfilesManager_UserWhitelistCollection]);
```

Comment:
You're right, our connection category is a real opportunity to make this code more readable.